File: packages/core/src/parser/__testdata__/sample.ex
defmodule Synapse.Agent do
  @moduledoc """
  Core agent module for Synapse.
  """

  use GenServer
  require Logger
  alias Synapse.Config
  import Synapse.Utils, only: [sanitize: 1, validate: 1]

  @max_retries 3
  @default_model "claude-opus-4-6"
  @timeout 30_000

  defstruct [:name, :config, status: :idle, tools: []]

  @type status :: :active | :idle | :stopped | :error
  @type t :: %__MODULE__{name: String.t(), config: Config.t(), status: status()}

  @callback process(message :: String.t()) :: {:ok, String.t()} | {:error, term()}
  @callback get_tools() :: [String.t()]

  def start_link(name, opts \\ []) do
    config = Keyword.get(opts, :config, %Config{})
    GenServer.start_link(__MODULE__, %{name: name, config: config}, name: via(name))
  end

  def process(agent, message) do
    GenServer.call(agent, {:process, message}, @timeout)
  end

  def get_status(agent) do
    GenServer.call(agent, :get_status)
  end

  @impl true
  def init(state) do
    Logger.info("Agent #{state.name} starting")
    {:ok, Map.put(state, :status, :idle)}
  end

  @impl true
  def handle_call({:process, message}, _from, state) do
    state = %{state | status: :active}
    result = call_model(message, state.config)
    state = %{state | status: :idle}
    {:reply, {:ok, result}, state}
  end

  @impl true
  def handle_call(:get_status, _from, state) do
    {:reply, state.status, state}
  end

  @impl true
  def handle_info(:timeout, state) do
    {:noreply, %{state | status: :idle}}
  end

  defp call_model(message, _config) do
    # TODO: implement actual API call
    "Response to: #{message}"
  end

  defp via(name), do: {:via, Registry, {Synapse.Registry, name}}

  defguard is_valid_message(msg) when is_binary(msg) and byte_size(msg) > 0

  defmacro __using__(_opts) do
    quote do
      import Synapse.Agent
    end
  end
end

defmodule Synapse.Config do
  defstruct model: "claude-opus-4-6", max_tokens: 4096, temperature: 0.7
end

defprotocol Synapse.Serializable do
  def serialize(data)
end

# FIXME: add supervision tree

File: packages/core/src/parser/__testdata__/sample.erl
-module(synapse_agent).
-behaviour(gen_server).

-export([start_link/2, process/2, get_tools/1, get_status/1]).
-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).

-define(MAX_RETRIES, 3).
-define(DEFAULT_MODEL, <<"claude-opus-4-6">>).
-define(TIMEOUT, 30000).

-record(config, {
    model = ?DEFAULT_MODEL :: binary(),
    max_tokens = 4096 :: non_neg_integer(),
    temperature = 0.7 :: float()
}).

-record(state, {
    name :: binary(),
    config :: #config{},
    status = idle :: active | idle | stopped | error
}).

-type status() :: active | idle | stopped | error.
-type agent_result() :: {ok, binary()} | {error, term()}.

-spec start_link(binary(), #config{} | undefined) -> {ok, pid()}.
start_link(Name, Config) ->
    gen_server:start_link({local, Name}, ?MODULE, [Name, Config], []).

-spec process(pid(), binary()) -> agent_result().
process(Agent, Message) ->
    gen_server:call(Agent, {process, Message}, ?TIMEOUT).

get_tools(_Agent) ->
    [<<"search">>, <<"read">>, <<"write">>].

get_status(Agent) ->
    gen_server:call(Agent, get_status).

init([Name, Config]) ->
    Cfg = case Config of
        undefined -> #config{};
        C -> C
    end,
    {ok, #state{name = Name, config = Cfg, status = idle}}.

handle_call({process, Message}, _From, State) ->
    case validate(Message) of
        true ->
            State1 = State#state{status = active},
            Result = call_model(Message, State1#state.config),
            State2 = State1#state{status = idle},
            {reply, {ok, Result}, State2};
        false ->
            {reply, {error, empty_message}, State}
    end;

handle_call(get_status, _From, State) ->
    {reply, State#state.status, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

call_model(Message, _Config) ->
    %% TODO: implement actual API call
    <<"Response to: ", Message/binary>>.

validate(Message) when is_binary(Message), byte_size(Message) > 0 -> true;
validate(_) -> false.

%% FIXME: add supervisor support
